bestie/utils: add C++ library to plumb google bench with bestie.

ccontavalli opened the pull request:
In short:
This commit provides a utility class and a few functions to allow
google bench library to save all metrics in a format that is
understood by bestie.

ccontavalli committed:
bestie/utils: add C++ library to plumb google bench with bestie.

In short:
This commit provides a utility class and a few functions to allow
google bench library to save all metrics in a format that is
understood by bestie.

jacob-adelmann commented on bestie/utils/googlebench/bestie_reporter_test.cc:
```diff
@@ -0,0 +1,735 @@
+  }
+}
+metrics {
+  metricname: "run__maxheapbytes_used"
```
Out of curiosity, why is this block replicated twice in kExpectedMessageText:

```
metrics {
  metricname: "run__maxheapbytes_used"
  tags {
    key: "run__benchmark_name"
    value: "Message"
  }
  tags {
    key: "run__skipped"
    value: "message"
  }
  tags {
    key: "run__skip_message"
    value: "For the powerful, crimes are those that others commit."
  }
  tags {
    key: "context__sys_info__name"
    value: "<hostname>"
  }
  tags {
    key: "context__executable_name"
    value: "<path-of-binary>"
  }
}
```

ccontavalli replied:
good catch, cleaned.

jacob-adelmann commented on bestie/utils/googlebench/bestie_reporter.cc:
```diff
@@ -0,0 +1,180 @@
+  context_.Clear();
+  context_.set_timestamp(epoch);
+
+  AddTag(&context_, "context__sys_info__name", context.sys_info.name);
```
Is context.sys_info.name the hostname the benchmark is run on?

jacob-adelmann replied:
I think yes based on details later in the test

jacob-adelmann commented on bestie/utils/googlebench/bestie_reporter.cc:
```diff
@@ -0,0 +1,180 @@
+  context_.set_timestamp(epoch);
+
+  AddTag(&context_, "context__sys_info__name", context.sys_info.name);
+  AddTag(&context_, "context__executable_name", context.executable_name);
```
Is context__executable_name just the executable name with the path stripped, or does it contain absolute or relative path?

justus-enf commented on bestie/utils/googlebench/bestie_reporter.cc:
```diff
@@ -0,0 +1,180 @@
+                             std::ios::out | std::ios::binary | std::ios::trunc | bestie_noreplace);
+    if (filestream.is_open()) return message.SerializeToOstream(&filestream);
+
+    if (errno != EEXIST) {
+      estream << "ERROR: saving benchmark results in " << filename
+              << " failed with errno: " << errno << "\n";
+      return false;
+    }
```
I think it is better to just check for file existance at the beginning rather than depending on the underlying implementation setting `errno`.

justus-enf replied:
Good point with the race. It was not clear to me that their was an exepctation of a race here.

jungleraptor reviewed (approved):
lgtm

ccontavalli commented:
/gcbrun although the failure does not seem to be related to this PR at all.

ccontavalli commented:
> /gcbrun although the failure does not seem to be related to this PR at all.

I stand corrected, I believe it's a missing dependency in the enkit repo. Basically: my PR is using googlebench for the first time in the enkit repo. googlebench depends on libpfm, but libpfm is not in the enkit repo -> failure. Will add a few commits to this PR once I figure it out.

https://github.com/google/benchmark/pull/1408/files (by @rajachan - thank you!!) added support for it upstream, there's probably a macro import that we're missing.

rajachan commented:
See https://github.com/google/benchmark/pull/1520, benchmark_deps() might be what you are looking for (I think).

ccontavalli committed:
bazel/init: explicit bring in google benchmark as a dependency.

Background:
In the last commit I added a library that depended on com_github_google_bench.
It built, it worked. Except CI/CD was failing on missing libpfm dependency.
On further inspection, it turns out that com_github_google_bench is not included
anywhere in our WORKSPACE file, or bzlmod equivalent file.

So we're getting google bench indirectly, via another dependency
(likely, grpc / protocol buffers are bringing it in).

In our internal/ repository, which uses enkit, we explicitly bring in
googlebench 1.5.1. Turns out that the rule there is ignored.
A simple build of a benchmark depending on google bench shows that we
are linking against version 1.9.0!!!

Which makes sense: enkit is imported first, in WORKSPACE file the
**first** definition wins. So the rule we have in our internal repository
is useless, and in enkit we are picking whichever random version is
brought in by 3rd party dependencies.

In this commit:
- explicitly bring in google bench 1.9.2
- invoke the macro that allows to bring in dependencies of google bench

Verified:
Built my library with google bench dependency. The version of google
bench shown is now 1.9.2 instead of 1.9.0.
Similar result when using --override_repository for enkit in our
internal repo.

After this PR is merged, I'll remove the redundant definitions.

ccontavalli commented:
(of course.. googlebench 1.9.0 vs 1.9.2, the content of the Run object changed, heap info is no longer an optional pointer, so now commits are 1. updating the library, 2. actually adding the glue between googlebench and bestie).

gpaussaenf reviewed (approved):
Approving so as to not hold up this PR. Others more familiar with the topic have already commented.
